### ISSUE TITLE
oc cluster up with --public-hostname 127.0.0.1 is still visible from outside of machine.

### ISSUE THREAD

**Issue by GrahamDumpleton** [openshift/origin#10628]
When running `oc cluster up` it uses the main IP address of the host rather than use `127.0.0.1`. This means it is accessible from outside of the host if there is no firewall set up. Because there is no effective user access controls as can supply any password, this means that anyone can get into the OpenShift instance and do stuff, with the possibility of launching enough applications to exhaust the host of resources.

Even when you use the `--public-hostname 127.0.0.1` option it still appears to be accessible from outside of the host and isn't restricting itself to only listening on the local loopback interface. When people access it from outside when set with this public hostname, the login redirect goes back to `127.0.0.1` of the client host. So web interface may be harder to get into but still may be a way. More than likely the REST API is still wide open though.

Using `--public-hostname localhost` instead doesn't help as it just results in public IP address of host being used and doesn't even use `127.0.0.1` as one might expect.

As to blocking access to external clients on MacOS X using a firewall, using the implied firewall access that MacOS X provides via System Preferences, you would need to block `socat` from being able to accept inbound requests. This means that you can't use `socat` for other uses where want inbound connections via it for other reasons.

What needs to be provided is a way of having it only be available from the host system itself via `127.0.0.1`, meaning it shouldn't be listening on external interfaces and only on `127.0.0.1`.

It may also be better if some way can be found for it to not be dependent on `socat`, but some OpenShift relay program with its own name so it alone can be blocked by firewall rules.

These observations were made on MacOS X, but expect that same applies to Linux as well.

**Comment by GrahamDumpleton** [comment]
Is the problem that in ``startSocatTunnel()`` it uses:

* https://github.com/openshift/origin/blob/eb9c389b604abb9c976ee0b09f42d2918713be79/pkg/bootstrap/docker/openshift/helper_unix.go#L66

```
	cmd := exec.Command("socat", "TCP-L:8443,reuseaddr,fork,backlog=20", "SYSTEM:\"docker exec -i origin socat - TCP\\:localhost\\:8443,nodelay\"")
```

Does the first argument to the ``socat`` run outside of the container need to ensure it uses the public hostname for the cluster, be that the generated IP if none supplied, or the supplied one. Thus if public hostname was ``127.0.0.1``, this would end up effectively being:

```
TCP-L:127.0.0.1:8443,reuseaddr,fork,backlog=20
```

That way the tunnel run outside of the Docker containers would only listen on interface for the IP.

I guess perhaps you only need do this though if was ``127.0.0.1`` or other IP explicitly given. If no public hostname given, then could still listen on all interfaces by using ``0.0.0.0`` of just leaving it out as is case now.

**Comment by GrahamDumpleton** [comment]
Port 1936 for router is also visible. It does take a password at least.

And I also know port 80/443 were visible so exposing your application if people knew host name. They would have to fudge things to have Host header sent. It was port 8443 was mainly concerned about as no security on accounts. All the ports exposed should be looked at though.

**Comment by csrwng** [comment]
It's harder to control 80/443 as those are exposed by Docker itself and as far as I know there is no way to tell Docker to bind to a specific interface. For the 8443 port, yes one option would be to forward traffic via an 'oc cluster' subcommand. One problem I see with binding to 127.0.0.1 is that when using xip.io for routes, containers running in OpenShift won't be able to access an application by route -- blah.127.0.0.1.xip.io will always resolve to the container itself and not to the router.

**Comment by GrahamDumpleton** [comment]
A container running in OpenShift shouldn't really be using the publicly accessible name. It should use the internal service name, skipping the router and going via the service IP local balancer.

**Comment by csrwng** [comment]
Yeah, I can't think of a good use case for it at the moment... just thinking that your routes would be limited in that way. They're also good only for your local machine, which I guess is perfectly ok if all you're doing is testing things locally.

**Comment by GrahamDumpleton** [comment]
As I talk about in another issue, the other way to approach it is to make it easier to set up the cluster with actual accounts and passwords user can choose so not open to everyone.

**Comment by csrwng** [comment]
I'll look into that as well.

**Comment by marekjelen** [comment]
+1 for adding some kind of security system rather than forcing 127.0.0.1. In case you want to work with cross-project services, you either need pod IP address (IMHO not recommended, as it would not work in other envs) or use the external domain name.

**Comment by csrwng** [comment]
Hi @GrahamDumpleton @marekjelen, so I think the simplest solution is to have socat limit the connection origin to 127.0.0.1 with the range option: http://www.dest-unreach.org/socat/doc/socat.html#GROUP_RANGE
PR forthcoming.

**Comment by GrahamDumpleton** [comment]
I am thinking I would prefer focus to be put on having a way to enable prompting for alternate developer password on startup the first time. Then it wouldn't matter so much that visible outside for case where using host IP rather than 127.0.0.1. If had prompt to change password, would happily use host IP.

**Comment by jawnsy** [comment]
@marekjelen @GrahamDumpleton Prompting for a password (or some method to set a password) sounds like it should be separate from this issue. I've created https://github.com/openshift/origin/issues/10919 for this. The underlying requirement is similar, but the solutions are tangential and we may want to use them separately.

**Comment by csrwng** [comment]
@GrahamDumpleton would you considered this fixed by https://github.com/openshift/origin/pull/11139 ?

**Comment by GrahamDumpleton** [comment]
Overall I am not sure if this is really resolved.

As per separate email discussion about it, you now have a situation where Linux behaves differently. So you have addressed a concern about MacOS X, but Linux still has the same vulnerability.

> MacOS X
>
> Public = 127.0.0.1 --> 8443 only listens on 127.0.0.1
> Public = 1.2.3.4 (public host IP) --> 8443 only listens on 127.0.0.1
>
> Linux
>
> Public = 127.0.0.1 --> 8443 only listens on 127.0.0.1
> Public = 1.2.3.4 (public host IP) --> 8443 listens on 1.2.3.4

The behaviour should really be consistent no matter what the platform.

It should also be enumerated what happens when using Docker machine vs direct Docker instance as well to see how things compare there.

So need to look at the bigger picture of this and not be happy that making a change over a concern about one specific platform.

**Comment by csrwng** [comment]
By necessity, I think things are going to be different depending on the platform. 

On OS X (or Windows), you will always need a VM to run Docker. If you're running the VM via Vagrant or Docker machine, then usually the VM will not be exposed to the external network. Of course, you can expose it if you really want to. If using Docker for Mac, with socat, we're simply now avoiding making port public by default.

Linux is a different case altogether because usually you're running Docker directly on the machine. If you want an experience similar to the Mac, you can always run it in a VM to isolate it.

**Comment by GrahamDumpleton** [comment]
So it will still be shipped as being vulnerable by default on Linux, but they will not be warned. I don't understand how that is seen as acceptable.

**Comment by csrwng** [comment]
Hmm, I don't know that we can enough about your environment to tell you whether you're vulnerable or not. When you run on a Mac or Windows, you are still running on a Linux machine/vm. 

It's also no different than running 'openshift start' on a Linux machine.

**Comment by csrwng** [comment]
I am closing this issue as the original was regarding access to the cluster from outside by default on the Mac. That is no longer the case.
